packages/web3-eth-abi/src/types.ts

```diff
@@ -1,57 +1,187 @@
+	readonly stateMutability: 'nonpayable' | 'payable';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+};
 
-export interface JsonAbiFunctionFragment extends JsonAbiBaseFragment {
-	type: 'function';
-	stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
-	outputs?: Array<JsonAbiParameter>;
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiFunctionFragment = AbiBaseFragment & {
+	readonly name: string;
+	readonly type: 'function';
+	readonly stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+	readonly outputs: ReadonlyArray<AbiParameter>;
 
 	// legacy properties
-	constant?: boolean;	// stateMutability == 'pure' or stateMutability == 'view'
-	payable?: boolean; // stateMutability == 'payable'
-}
+	readonly constant?: boolean; // stateMutability == 'pure' or stateMutability == 'view'
+	readonly payable?: boolean; // stateMutability == 'payable'
+};
 
-export interface JsonAbiEventFragment extends JsonAbiBaseFragment {
-	type: 'event';
-	anonymous?: boolean;
-}
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiEventFragment = AbiBaseFragment & {
+	readonly name: string;
+	readonly type: 'event';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+	readonly anonymous?: boolean;
+};
 
 // https://docs.soliditylang.org/en/latest/abi-spec.html#json
-export type JsonAbiFragment =
-	| JsonAbiConstructorFragment
-	| JsonAbiFunctionFragment
-	| JsonAbiEventFragment;
+export type AbiFragment = AbiConstructorFragment | AbiFunctionFragment | AbiEventFragment;
+
+export type ContractAbi = ReadonlyArray<AbiFragment>;
+
+export type AbiInput = string | AbiParameter | { readonly [key: string]: unknown };
+
+export type FilterAbis<Abis extends ContractAbi, Filter, Abi = Abis[number]> = Abi extends Filter
+	? Abi
+	: never;
+
+type _TypedArray<Type, Size extends string> = Size extends ''
+	? Type[]
+	: FixedSizeArray<Type, ConvertToNumber<Size>>;
+
+export type PrimitiveAddressType<Type extends string> = Type extends `address[${infer Size}]`
+	? _TypedArray<Address, Size>
+	: Type extends 'address'
+	? Address
+	: never;
+
+export type PrimitiveStringType<Type extends string> = Type extends `string${string}[${infer Size}]`
+	? _TypedArray<string, Size>
+	: Type extends 'string' | `string${string}`
+	? string
+	: never;
+
+export type PrimitiveBooleanType<Type extends string> = Type extends `bool[${infer Size}]`
+	? _TypedArray<boolean, Size>
+	: Type extends 'bool'
+	? boolean
+	: never;
+
+export type PrimitiveIntegerType<Type extends string> = Type extends `uint${string}[${infer Size}]`
+	? _TypedArray<Numbers, Size>
+	: Type extends 'uint' | 'int' | `int${string}` | `uint${string}`
+	? Numbers
+	: never;
+
+export type PrimitiveBytesType<Type extends string> = Type extends `bytes${string}[${infer Size}]`
+	? _TypedArray<Bytes, Size>
+	: Type extends 'bytes' | `bytes${string}`
+	? Bytes
+	: never;
+
+export type PrimitiveTupleType<
```

Comment:
also as discussed could you see if complex types checking like mappings , mappings of structs or mappings of mappings support is applicable for this feature?

Reply:
As per my knowledge mapping and mapping of structures are only be used internally by solidity. Can't be used as input or output of a public method or event.